### ISSUE TITLE
NaN needs to be handled here, otherwise JSON.stringfly will convert NaN to null below

### ISSUE THREAD

**Issue by mongolming** [log4js-node/log4js-node#1187]
`serialise() {
    const logData = this.data.map((e) => {
      // JSON.stringify(new Error('test')) returns {}, which is not really useful for us.
      // The following allows us to serialize errors correctly.
      if (e && e.message && e.stack) {
        e = Object.assign({ message: e.message, stack: e.stack }, e);
      }
      if(isNaN(e) && typeof e === 'number') {
        e = 'NaN'
      }
      return e;
    });
    this.data = logData;
    return flatted.stringify(this);
  } `

**Comment by peteriman** [comment]
Hi @mongolming,

1. Do you have the test code/case for it with the expected and actual values? (to understand how this will happen)
2. Will it impact the `deserialise()` function? (regression test)
3. Would you mind submitting a PR for it? (so you can be properly credited)
 You should be able to edit the code of `LoggingEvent.js` or you can also [click here](https://github.com/log4js-node/log4js-node/edit/master/lib/LoggingEvent.js) to proceed.

**Comment by peteriman** [comment]
I did some preliminary checks.
@mongolming, you might want to continue from here.

```diff
   serialise() {
     const logData = this.data.map((e) => {
       // JSON.stringify(new Error('test')) returns {}, which is not really useful for us.
       // The following allows us to serialize errors correctly.
       if (e && e.message && e.stack) {
         e = Object.assign({ message: e.message, stack: e.stack }, e);
       }
+      if (typeof e === 'number' && (isNaN(e) || !isFinite(e))) {
+        e = e.toString();
+      }
       return e;
     });
     this.data = logData;
     return flatted.stringify(this);
   }
```

**Comment by mongolming** [comment]
or
`return flatted.stringify(this, (key,value) => typeof value === 'number' && (isNaN(value) || !isFinite(value))
 ? value.toString() 
 : value ;`

**Comment by mongolming** [comment]
@peteriman Whether to consider whether the multi-layer data contains NaN, undefined

**Comment by peteriman** [comment]
Added multi-layer data as a task in PR #1188.

**Comment by peteriman** [comment]
Only affects clustering, multiprocessAppender, and tcpAppender.
These three will `serialise()` to `String` to transmit for the receiver to `deserialise()`.

Due to `JSON.stringify()`, there are some translation loss:
| Code | Object | JSON.stringify() | Fix |
|-|-|-|-|
`{"a": parseInt("abc")}` | `{"a": NaN}` | `{"a": null}` | `{"a": "NaN"}` | 
`{"b": 1/0}` | `{"b": Infinity}` | `{"b": null}` | `{"c": "Infinity"}` |
`{"c": -1/0}` | `{"c": -Infinity}` | `{"c": null}` | `{"c": "-Infinity"}` |
`[undefined]` | `[undefined]` | `[null]` | `["undefined"]` |

While the fix parses them to `String` instead of the `Object`, it does not affect the output in log files.

Fixed in `log4js@^6.4.2`.

For more details: https://github.com/log4js-node/log4js-node/pull/1188#issue-1142624744

**Comment by peteriman** [comment]
Fixed in `log4js@^6.4.2`.

For more details: https://github.com/log4js-node/log4js-node/pull/1188#issue-1142624744

**Comment by peteriman** [comment]
Released `log4js@6.4.2`.

@mongolming Please let me know if there are still issues after upgrading.

**Comment by lamweili** [comment]
## Affected Components
Only affects clustering, multiprocessAppender, and tcpAppender.
These three will `serialise()` to `String` to transmit for the receiver to `deserialise()`.

| Code | Object<br>(Input) | Serialised<br>(Transmission) | Deserialised<br>(Output) | Match
|-|-|-|-|-|
`{"a": Number("abc")}` | `{"a": NaN}` | `{"a": "__LOG4JS_NaN__"}` | `{"a": NaN}` | ✔️ |
`{"b": 1/0}` | `{"b": Infinity}` |  `{"b": "__LOG4JS_Infinity__"}` | `{"b": Infinity}` | ✔️ |
`{"c": -1/0}` | `{"c": -Infinity}` | `{"c": "__LOG4JS_-Infinity__"}` | `{"c": -Infinity}` | ✔️ |
`[undefined]` | `[undefined]` | `["__LOG4JS_undefined__"]` | `[undefined]` | ✔️ |

Updated by PR #1332.
